Show a per-layer configurable color while a layer is active

ctranstrum opened the pull request:
I installed the rgbled widget on a xiao rp2040, and was a bit disappointed that ... nothing happened.

So I thought to fork the repo and make a rp2040-specific version that lets you define a color for each layer. But then as I was developing it, I realized there's not really a reason this couldn't be part of the original repo, since it's only active while a layer is active, and the rest of the time it's off.

But I'm getting ahead of myself. Summary of feature:

- by default, nothing, so that this doesn't impact anyone currently using the module.
- if enabled, turns on the light when a layer is active.
- the color for each layer can be configured.
- when a layer is set to black (which is the default for the base layer) the led is off, and the behavior is the same as without this option.

Up to you if you want to accept this pull request, if not, I can maintain this fork.

caksoylar commented:
It looks good to me, very nicely implemented as well, thanks! One thing I am wondering is if a color is active and you want to show a blink of the same color, is there a period of lights-off around the blink so it can be noticed? It didn't seem to me from a cursory look at the code and it would be nice to have that if it doesn't complicate the implementation too much.

As an aside, I think we are hitting the limits of Kconfig-based configuration. Things need to be migrated to use DT nodes for settings eventually but given it would break things anyway, that's something to do in a 2.0 or a replacement module.

ctranstrum commented:
Thanks, yeah, that's a good idea for the separation of a same-color blink. I can add that.

I also noticed that the `first_item` attribute of the `blink_item` struct doesn't seem to be used, but I didn't want to remove it if there is some secret reason it exists.

DT nodes sounds like a good idea as well, but I'll leave that for you in the 2.0 release.

caksoylar commented:
I think the `first_item` is a vestige from an older implementation that got refactored; feel free to remove it if you are touching around it.

ctranstrum commented:
Okay, removed the `first_item` and added a "black blink" if the blink is the same color as the current led color. Tested on my board and everything seems to work.

ctranstrum commented:
Oh, just had a thought: I don't think this will work correctly with the option to show the layer change option using the series of blinks. If you enable both color layers as well as blinky-layer-change, then when you switch to a layer that is the same color as your layer blink color, then it'll look like you have double the number of your current layer, as the separation between blinks will be doubled up and it will go back to the layer color in between blinks.

Not sure why anyone would enable both options at once, but we can either put a note in to choose only one of the two layer options or I can add some extra code for that one rather rare use case. What do you think?

caksoylar commented:
Good thinking -- can you add a build assert to make sure both aren't turned on at once? Let me know if you need a pointer.

ctranstrum commented:
Will do. I suspect you can tell by the nature and number of my commits that C isn't my most proficient language. :)

But it looks like you already have some build asserts in the source, so it should be pretty straight-forward to add another. I'll take a stab at it this evening.

ctranstrum commented:
Tested the build assert on my board and got a failure as expected when I enabled both options, and a successful build with only one option.

caksoylar commented on src/widget.c:
```diff
@@ -161,6 +184,37 @@ ZMK_LISTENER(led_battery_listener, led_battery_listener_cb);
+        led_current_layer_color = layer_color_idx[index];
+        struct blink_item color = {.duration_ms = 5, .color = led_current_layer_color, .sleep_ms = 5};
+        LOG_INF("Setting layer color to %s for layer %d", color_names[led_current_layer_color], index);
+        k_msgq_put(&led_msgq, &color, K_NO_WAIT);
```
I think this doesn't do much besides making the thread run again and update with the current color, right?

ctranstrum replied:
Because I'm not an expert in C or Zephyr, I was afraid to control the LED devices from two different threads, so anything that needs to update the leds, I put into the led_msgq. If we don't have to worry about threads and etc, we could just update the LEDs from here, since we aren't performing a blink. By using the queue, though, we can also interact more cleanly with any concurrent blinks.

caksoylar replied:
That's fair, depending on the thread is a good idea, yes. I am just wondering if we should have a "dummy" blink type, where it doesn't do anything but only updates the color. Maybe if `blink_ms` is zero, what do you think?

ctranstrum replied:
Yes, that's essentially what I did, using the existing mechanisms and setting the duration_ms and sleep_ms to 5ms each, which is as close to 0 as I was comfortable getting without knowing if there were any downsides to rapidly toggling the led.

There is existing code that checks for sleep_ms being omitted, and we could add something similar for duration_ms being omitted, and if so, skip the wait state. That way we wouldn't have to add a new attribute to the struct.

ctranstrum replied:
I think that was the right call. Once I started to separate the different paths, the logic became much cleaner, and I was able to refactor to make it much more obvious what is going on.

caksoylar commented on src/widget.c:
```diff
@@ -161,6 +184,37 @@ ZMK_LISTENER(led_battery_listener, led_battery_listener_cb);
+static int led_layer_color_listener_cb(const zmk_event_t *eh) {
+    ARG_UNUSED(eh);
+
+    if (!initialized) {
+      return 0;
+    }
+
+    update_layer_color();
+
+    return 0;
```
```suggestion
    if (initialized) {
        update_layer_color();
    }
    return 0;
```
Cosmetic change

caksoylar commented on src/widget.c:
```diff
@@ -216,20 +286,28 @@ extern void led_process_thread(void *d0, void *d1, void *d2) {
-            }
+        if (blink.color == current_color && current_color != 0) {
+            set_rgb_leds(blink.color, 0);
+            k_sleep(K_MSEC(blink.duration_ms));
```
Maybe we can use this value instead?
```suggestion
            k_sleep(K_MSEC(CONFIG_RGBLED_WIDGET_INTERVAL_MS));
```

caksoylar reviewed (commented):
Did you try this with conditional layers, out of curiosity? I am wondering if it could benefit from debouncing like used with the layer blinks.

ctranstrum commented:
Yes, I have conditional layers on my board, and I had an update because it wasn't very responsive at first, but I realized it was because of the default CONFIG_RGBLED_WIDGET_INTERVAL_MS setting, and adding a layer-specific .sleep_ms solved it.

To be fair, I don't really understand the debounce stuff, so if you feel it's still better to use that, I will defer to your judgment on that.

caksoylar commented:
> Yes, I have conditional layers on my board, and I had an update because it wasn't very responsive at first, but I realized it was because of the default CONFIG_RGBLED_WIDGET_INTERVAL_MS setting, and adding a layer-specific .sleep_ms solved it.
> 
> To be fair, I don't really understand the debounce stuff, so if you feel it's still better to use that, I will defer to your judgment on that.

If the current state works fine for you, I'd be OK with merging this after I get a chance to test it myself. I can then see if incorporating debouncing is better or not.

ctranstrum commented:
Just installed the latest on my board, and I see no issues. Let me know if you have any thoughts regarding my latest refactor. I think the code is a lot cleaner now thanks to your suggestions.

caksoylar commented on src/widget.c:
```diff
@@ -211,32 +282,32 @@ extern void led_process_thread(void *d0, void *d1, void *d2) {
+            if (separation) {
+                set_rgb_leds(0, CONFIG_RGBLED_WIDGET_INTERVAL_MS);
+            }
+            set_rgb_leds(previous_color, blink.sleep_ms);
```
Looks good to me, thanks for the refactoring! One final nit, because it initially confused me trying to understand it and this seems equivalent (but easier to grok) to me, right?

```suggestion
            set_rgb_leds(previous_color, blink.sleep_ms > 0 ? blink.sleep_ms : CONFIG_RGBLED_WIDGET_INTERVAL_MS);
```
(I can't create a suggestion that includes lines below, but this would also remove lines 305-307, and we can move the comment on line 304 to before 302)

ctranstrum replied:
Ah, yes, excellent idea! Thanks!

ctranstrum commented:
Just tested after updating with your latest suggestion, and everything looks good. Thanks so much for your patience with me on this.

caksoylar commented:
I just got around to testing this, thank *you* for the very nice contribution!
